knora/dsplib/utils/onto_create_lists.py

```diff
@@ -1,141 +1,165 @@
+from .shared_methods import login, try_network_action
 
 
-def create_list_node(con: Connection, project: Project, parent_node: ListNode, nodes: list[dict[str, Any]]) -> list[
-    dict[str, Any]]:
+def _create_list_node(
+    con: Connection,
+    project: Project,
+    node: dict[str, Any],
+    parent_node: Optional[ListNode] = None
+) -> Tuple[dict[str, Any], bool]:
     """
-    Creates the list on the DSP server
+    Creates a list node on the DSP server, recursively scanning through all its subnodes, creating them as well.
+    Returns a tuple consisting of a dict and a bool. The dict contains the IRIs of the created list nodes. The bool
+    indicates if all nodes could be created or not.
 
     Args:
-        con: Connection to the DSP server
-        project: Project which the lists should be added to
-        parent_node: Root node of the list
-        nodes: List of nodes the list is made of
+        con: connection to the DSP server
+        project: project that holds the list where this node should be added to
+        node: the node to be created
+        parent_node: parent node of the node to be created (optional)
 
     Returns:
-        The list of all nodes with their names and respective IDs
+        dict of the form ``{nodename: {"id": node IRI, "nodes": {...}}}`` with the created list nodes, nested according to their hierarchy structure
+        True if all nodes could be created, False if any node could not be created
     """
-    nodelist = []
-    for node in nodes:
-        new_node = None
-        try:
-            new_node = ListNode(con=con, project=project, label=node["labels"], comments=node.get("comments"),
-                                name=node["name"],
-                                parent=parent_node).create()
-        except BaseError as err:
-            print(f"ERROR while trying to create list node '{node['name']}'. The error message was {err.message}")
-            exit(1)
-        if new_node:
-            # if node has child nodes, call the method recursively
-            if node.get("nodes"):
-                subnode_list = create_list_node(con, project, new_node, node["nodes"])
-                nodelist.append({new_node.name: {"id": new_node.id, "nodes": subnode_list}})
-            else:
-                nodelist.append({new_node.name: {"id": new_node.id}})
-    return nodelist
-
-
-def create_lists(input_file: str, lists_file: str, server: str, user: str, password: str, verbose: bool,
-                 dump: bool = False) -> dict[str, Any]:
+    new_node = ListNode(
+        con=con,
+        project=project,
+        label=node["labels"],
+        comments=node.get("comments"),
+        name=node["name"],
+        parent=parent_node
+    )
+    try:
+        new_node = try_network_action(
+            object=new_node,
+            method="create",
+            error_message_on_failure=f"ERROR while trying to create list node '{node['name']}'."
+        )
```

Comment:
I think this method would be much cleaner if it would always accept only a callable. And in instances like this, where you can't simply pass a function to it, you'd be using a lambda.  
So this would look as follows:

```python
new_node = try_network_action(lambda: new_node.create(), f"ERROR ...")
```

Reply:
Thank you very much, this is a great improvement! I tried it out, and it works :-) You always have to use a lambda, so that the method is not executed too early. This is really a big improvement!

Reply:
cool :)